File: TurtleFlight/Core/Flight/FlightEngine.swift
import Foundation
import SceneKit

final class FlightEngine {
    // MARK: - Flight State
    struct FlightState {
        var position: SCNVector3 = SCNVector3(0, 500, 0)
        var rotation: SCNVector3 = .init(0, 0, 0) // euler angles
        var heading: Float = 0       // degrees, 0 = North
        var speed: Float = 200       // km/h
        var altitude: Float = 500    // meters
        var verticalSpeed: Float = 0 // m/s
        var isBoosting: Bool = false
        var boostTimeRemaining: Float = 0
        /// Boost gauge in [0, 1]. A boost needs a full gauge and drains it;
        /// stars refill it (`registerStarCollected`) and it trickles back
        /// on its own. Starts full so the first flight can boost at once.
        var boostCharge: Float = 1
        /// True while the energy model has the character stalled (Expert
        /// only): too slow to hold altitude, nose drops until speed returns.
        var isStalled: Bool = false
        var flightTime: TimeInterval = 0
    }

    // MARK: - Properties
    private(set) var state = FlightState()
    private var profile: SensitivityProfile
    /// Per-vehicle multipliers applied on top of the sensitivity profile
    /// each frame. Defaults to `.neutral` (all 1.0) for unit tests and
    /// pre-vehicle-selection state. FlightViewModel sets this in
    /// `startFlight(...)` to the current vehicle's handling.
    var vehicleHandling: VehicleHandling = .neutral

    /// Un-boosted cruise speed the energy model relaxes toward (km/h).
    /// Separate from `state.speed` so boost can be layered on top as an
    /// instantaneous multiplier without feeding back into the model.
    private var energySpeed: Float = Constants.Flight.defaultSpeed

    // MARK: - Init
    init(sensitivity: SensitivityLevel = .easy) {
        self.profile = SensitivityProfile.profile(for: sensitivity)
    }

    // MARK: - Public Methods

    func updateSensitivity(_ level: SensitivityLevel) {
        profile = SensitivityProfile.profile(for: level)
    }

    func reset() {
        state = FlightState()
        energySpeed = Constants.Flight.defaultSpeed
    }

    /// A collected star tops up the boost gauge.
    func registerStarCollected() {
        state.boostCharge = min(1, state.boostCharge + Constants.Flight.boostChargePerStar)
    }

    /// True when tapping Boost right now would start one.
    var canBoost: Bool {
        !state.isBoosting && state.boostCharge >= 1
    }

    /// Main update loop - call every frame
    func update(
        deltaTime: Float,
        rollInput: Double,
        pitchInput: Double,
        isBoosting: Bool,
        shouldAutoLevel: Bool
    ) {
        state.flightTime += Double(deltaTime)

        // Boost: needs a full gauge, drains it, then the gauge trickles
        // back (and jumps on every star) so the player has a reason to
        // chase stars and a rhythm to when they can punch it again.
        if isBoosting && state.boostTimeRemaining <= 0 && state.boostCharge >= 1 {
            state.isBoosting = true
            state.boostTimeRemaining = Float(Constants.Flight.boostDuration)
            state.boostCharge = 0
        }
        if state.boostTimeRemaining > 0 {
            state.boostTimeRemaining -= deltaTime
            if state.boostTimeRemaining <= 0 {
                state.isBoosting = false
            }
        } else {
            state.boostCharge = min(1, state.boostCharge
                                    + Constants.Flight.boostRechargePerSecond * deltaTime)
        }

        let speedMultiplier: Float = state.isBoosting ? Constants.Flight.boostMultiplier : 1.0

        // Vertical speed from pitch input, with per-vehicle pitch multiplier
        // (e.g. balloon body has the highest lift at 1.15).
        let pitchRate = Float(profile.pitchSpeed) * vehicleHandling.pitch * Float(pitchInput)
        state.verticalSpeed = pitchRate

        // Auto-level
        if shouldAutoLevel {
            state.verticalSpeed *= 0.95 // Gradually reduce
        }

        // Energy model (Normal / Expert): diving trades altitude for
        // speed, climbing bleeds it, and drag pulls the speed back toward
        // cruise. Easy keeps the flat constant-speed model so a first
        // flight never has to think about airspeed.
        if profile.energyModelEnabled {
            let base = Constants.Flight.defaultSpeed
            let energyGain = -state.verticalSpeed * Constants.Flight.energyExchangeRate
            let drag = (energySpeed - base) * Constants.Flight.energyDrag
            energySpeed += (energyGain - drag) * deltaTime
            energySpeed = energySpeed.clamped(to: Constants.Flight.minEnergySpeed...Constants.Flight.maxEnergySpeed)
        } else {
            energySpeed = Constants.Flight.defaultSpeed
        }

        // Stall (Expert): below the stall speed the nose drops regardless
        // of input and turning authority collapses until speed recovers.
        let stalled = profile.stallEnabled && energySpeed < Constants.Flight.stallSpeed
        state.isStalled = stalled
        if stalled {
            state.verticalSpeed = min(state.verticalSpeed, -Constants.Flight.stallSinkRate)
        }

        state.speed = energySpeed * speedMultiplier

        // Heading (yaw) from roll input, scaled by the current vehicle's
        // turn multiplier so e.g. the carrot jet (1.10) feels noticeably
        // sharper than the cushion balloon (0.85).
        let turnAuthority: Float = stalled ? 0.35 : 1.0
        let turnRate = Float(profile.turnSpeed) * vehicleHandling.turn * Float(rollInput) * deltaTime * turnAuthority
        state.heading += turnRate
        if state.heading >= 360 { state.heading -= 360 }
        if state.heading < 0 { state.heading += 360 }

        // Update altitude
        state.altitude += state.verticalSpeed * deltaTime

        // Min altitude protection
        let minAlt = Float(profile.minAltitude)
        if state.altitude < minAlt {
            state.altitude = minAlt
            state.verticalSpeed = max(state.verticalSpeed, 0)
        }

        // Max altitude
        if state.altitude > Constants.Flight.maxAltitude {
            state.altitude = Constants.Flight.maxAltitude
            state.verticalSpeed = min(state.verticalSpeed, 0)
        }

        // Update position based on heading and speed
        let headingRad = state.heading.rad
        let speedMS = state.speed / 3.6 * deltaTime // km/h to m/s
        let dx = sin(headingRad) * speedMS
        let dz = -cos(headingRad) * speedMS

        state.position.x += dx
        state.position.y = state.altitude
        state.position.z += dz

        // Authoritative euler for the character node.
        // CharacterAnimator may layer additional per-vehicle motion on top
        // of these axes (additive on .x for bellyGlider, override on .z
        // for cloudSurf, etc.) but no longer recomputes the base bank/pitch.
        // Bank carries the per-vehicle bias so a carrot-jet hard turn reads
        // as committed and a cushion-balloon turn stays politely level.
        let bankAngle = -Float(rollInput) * Constants.Camera.bankingAngle * 2 * vehicleHandling.bank
        let pitchAngle = Float(pitchInput) * 0.2
        state.rotation = SCNVector3(pitchAngle, -headingRad, bankAngle)
    }

    /// Stall check
    var isStalling: Bool {
        guard profile.stallEnabled else { return false }
        return state.isStalled
    }
}

File: src/core/flight/FlightEngine.ts
import { SensitivityConfig } from '../gyro/SensitivityProfile';

export interface FlightState {
  position: { x: number; y: number; z: number };
  rotation: { yaw: number; pitch: number; roll: number };
  speed: number;
  altitude: number;
  heading: number;
  flightTime: number;
  isBoosting: boolean;
  starCount: number;
}

export class FlightEngine {
  private state: FlightState;
  private baseSpeed = 200;
  private boostMultiplier = 1.5;
  private maxSpeed = 500;

  constructor() {
    this.state = this.createInitialState();
  }

  private createInitialState(): FlightState {
    return {
      position: { x: 0, y: 100, z: 0 },
      rotation: { yaw: 0, pitch: 0, roll: 0 },
      speed: 200,
      altitude: 100,
      heading: 0,
      flightTime: 0,
      isBoosting: false,
      starCount: 0,
    };
  }

  update(dt: number, gyroInput: { roll: number; pitch: number }, profile: SensitivityConfig): FlightState {
    // 1. Update heading from roll input
    const turnRate = gyroInput.roll * profile.turnSpeed;
    this.state.heading = (this.state.heading + turnRate * dt + 360) % 360;
    this.state.rotation.yaw = this.state.heading;
    this.state.rotation.roll = gyroInput.roll * 30; // visual bank angle

    // 2. Update pitch
    const pitchRate = gyroInput.pitch * profile.pitchSpeed;
    this.state.rotation.pitch = gyroInput.pitch * 20;

    // 3. Speed calculation
    let targetSpeed = this.baseSpeed;
    if (this.state.isBoosting) {
      targetSpeed = this.baseSpeed * this.boostMultiplier;
    }
    // Stall check
    if (profile.stallEnabled && this.state.speed < profile.stallSpeed) {
      targetSpeed = Math.max(targetSpeed * 0.5, 50);
    }
    this.state.speed += (targetSpeed - this.state.speed) * 2.0 * dt;
    this.state.speed = Math.min(this.state.speed, this.maxSpeed);

    // 4. Position update
    const headingRad = (this.state.heading * Math.PI) / 180;
    const speedMs = this.state.speed / 3.6; // km/h to m/s
    this.state.position.x += Math.sin(headingRad) * speedMs * dt;
    this.state.position.z += Math.cos(headingRad) * speedMs * dt;
    this.state.position.y += pitchRate * dt;

    // 5. Altitude protection
    if (this.state.position.y < profile.minAltitudeProtection) {
      this.state.position.y = Math.max(this.state.position.y, profile.minAltitudeProtection);
    }
    // Max altitude cap
    if (this.state.position.y > 2000) {
      this.state.position.y = 2000;
    }
    this.state.altitude = this.state.position.y;

    // 6. Flight time
    this.state.flightTime += dt;

    return { ...this.state };
  }

  setBoost(active: boolean): void {
    this.state.isBoosting = active;
  }

  collectStar(): void {
    this.state.starCount++;
  }

  getState(): FlightState {
    return { ...this.state };
  }

  reset(): void {
    this.state = this.createInitialState();
  }
}
